README
Test-WWW-Mechanize

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it get an idea of the modules uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2004 Andy Lester

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Test-WWW-Mechanize

WWW::Mechanize and Test::WWW::Mechanize do not use rt.cpan.org for
bug tracking.  They are now being tracked via GitHub at
https://github.com/petdance/test-www-mechanize/issues


1.60 Sun Dec  4 10:18:28 PM CST 2022
------------------------------------
[ENHANCMENTS]
Added default test messages for button_exists_ok and lacks_button_ok
methods.  Thanks, Daniel Böhmer (GH#70)

get_ok(), post_ok(), head_ok(), put_ok() and delete_ok() now all fail
if an undef URL is passed, rather than passing the undef URL into the
underlying LWP functions. Thanks, Jess Robinson. (GH #40)

[FIXES]
Disable proxies in some of the tests that look at a local test server.
(GH#55, GH#74)

[DOCUMENTATION]
Clarified some wording.  Thanks, Daniel Böhmer (GH#70)


1.58    Fri Apr 29 11:23:39 CDT 2022
------------------------------------
[ENHANCEMENTS]
Failed get_ok, post_ok, head_ok, put_ok, delete_ok calls now show the URL
as a diagnostic if there's a problem accessing the URL.  This apply to
functions like submit_form_ok() and follow_link_ok() that are wrappers
around these methods.


1.56    Tue Apr 26 22:14:12 CDT 2022
------------------------------------
This version of Test::WWW::Mechanize requires Perl 5.10.

[ENHANCEMENTS]
Add check_all_images_ok(). (GH #17)  Thanks, Julien Fiegehenn.
This is the first version of this function, and its functionality may
change in the near future.

[DOCUMENTATION]
Fixed some incorrect docs. Thanks, Daniel Böhmer.

[INTERNALS]
Require Perl 5.10.

Explicitly requires HTTP::Message 6.29 or higher.


1.54    Tue Dec  8 23:25:06 CST 2020
------------------------------------
[ENHANCEMENTS]
Use ok() instead of cmp_ok() inside of lacks_uncapped_inputs().
This output makes more sense.

lacks_uncapped_inputs() now has a a default message if one isn't supplied.

[FIXES]
Fixed the subtest name inside of C<lacks_ids_ok>.

Fixed the minimum version of Carp::Assert::More in Makefile.PL.


1.52    Wed Dec  5 10:00:56 CST 2018
------------------------------------
[ENHANCEMENTS]
click_ok() method can now take a button with X/Y coordinates.
Thanks to GitHub user @marderh.  (GH #45)

Added the ability to modify the HTML that the html_tidy_ok() validates.
See the content_for_validation() method. (GH #61)

Add a set of methods for existence of IDs: id_exists(), id_exists_ok(),
ids_exist_ok(), lacks_id_ok() and lacks_ids_ok(). (GH #48)

Add button_exists(), button_exists_ok() and lacks_button_ok()
methods.  (GH #50)

[FIXES]
Clarified warnings to make it clear that certain modules are
optional. Thanks, Matthew Chae.


1.50    Mon Apr 16 15:16:59 CDT 2018
------------------------------------
[FIXES]
Added html_tidy_ok() methods, analogous to html_lint_ok().

Remove unnecessary dependency on HTML::TreeBuilder.  Thanks, Kent Fredric.


1.49_01 Mon Mar 26 10:58:51 CDT 2018
------------------------------------
[ENHANCEMENTS]
Adding autotidy functionality. autotidy lets you validate every page that
Mech gets using the HTML::Tidy5 module, just like the autolint feature
does with the HTML::Lint module. HTML::Tidy5 is a much more complete
HTML checking tool, and validates HTML5 which HTML::Lint does not. You
must have HTML::Tidy5 1.00 installed to use autotidy.


1.48    Thu Dec 29 22:45:29 CST 2016
------------------------------------
[FIXES]
The linting method html_lint_ok() was not calling the HTML::Lint API
correctly, so may have missed some HTML errors at the end of a page.
This also applies to get, post, etc if you have the autolint argument on.


1.46
------------------------------------
[ENHANCEMENTS]
Added header_exists_ok(), lacks_header(), header_is() and header_like()
methods.  Thanks to Eric A. Zarko for the original patches.

The scraped_id_is() method used to assign a description for the test if
one was not passed.  Now it does not.

scraped_id_is() now gives proper diagnostics if an ID is not found in
the HTML.

Added a delete_ok() method.  Thanks, moznion.

content_contains() now fails if it's called with a regex.  content_like()
now fails if it's not called with a regex.

[FIXES]
The test server run during the test suite allowed URLs outside of the
document tree, which could potentially be a security problem.  This has
been fixed.  Thanks, Tynovsky.
https://github.com/petdance/test-www-mechanize/issues/33

Fixed an overly-restrictive optimization in scrape_text_by_id(), plus
scraped_id_is() and scraped_id_like() which wrap it.

The method checks to make sure that it doesn't bother looking for an ID
on the page if the ID doesn't exist.  It did this by looking for the text

    id="foo"

where foo is the ID being searched for.  However, that would mean that
tags like

    <p id='foo'>
    <p id=foo>
    <p id = "foo">
    <p id=
        "foo">
    <p id
    =
            "foo"
                    >

would be seen as not existing.  This has been fixed by making
scrape_text_by_id() search for the string "foo" anywhere on the page.


1.44    Sat Jun 30 20:32:04 CDT 2012
------------------------------------
There is no new functionality in this release.

[FIXES]
Fixed test failures on Win32.  Thanks, Jerry Gay.


1.42    Thu May 31 11:35:26 CDT 2012
------------------------------------
If you want to use the autolint functionality, you'll have to have
HTML::Lint 2.20.

[FIXES]
Custom lint objects don't get reset before they get used, making
autolint with a custom lint object practically useless.
https://github.com/petdance/test-www-mechanize/issues/25


1.40    Fri Apr 13 15:14:39 CDT 2012
------------------------------------
[ENHANCEMENTS]
Added a $mech->autolint() method so you can turn autolinting on and off
in mid-program.  Thanks, Mike O'Regan.

New functions $mech->scrape_text_by_id() and $mech->scrape_text_by_attr()
let you extract text from your pages.  Convenience function
$mech->scraped_id_is() makes it easy to scrape and compare in one
function.

    <h1 id="pagetitle">My Awesome Page!</h1>

    # Verify that HTML is there with:

t/HangServer.pm
package HangServer;

use warnings;
use strict;

BEGIN {
    delete $ENV{http_proxy}; # All our tests are running on localhost
}

use base 'HTTP::Server::Simple::CGI';

use Carp ();

our $pid;

sub new {
    my $class = shift;

    die 'An instance of HangServer has already been started.' if $pid;

    # XXX This should really be a random port.
    return $class->SUPER::new(13432, @_);
}

sub run {
    my $self = shift;

    $pid = $self->SUPER::run(@_);

    $SIG{__DIE__} = \&stop;

    return $pid;
}

sub handle_request {
    my $self = shift;
    my $cgi  = shift;

    use Data::Dumper; local $Data::Dumper::Sortkeys=1;
    my $response = Dumper($cgi);
    print "HTTP/1.0 200 OK\r\n";
    print "Content-Type: text/html\r\nContent-Length: ", length($response), "\r\n\r\n", $response;
    return;

    my $file = (split( /\//,$cgi->path_info))[-1]||'index.html';
    $file    =~ s/\s+//g;

    my $filename = "t/html/$file";
    if ( -r $filename ) {
        if (my $response=do { local (@ARGV, $/) = $filename; <> }) {
            print "HTTP/1.0 200 OK\r\n";
            print "Content-Type: text/html\r\nContent-Length: ", length($response), "\r\n\r\n", $response;
            return;
        }
    }

    print "HTTP/1.0 404 Not Found\r\n\r\n";

    return;
}

sub background {
    my $self = shift;

    $pid = $self->SUPER::background()
        or Carp::confess( q{Can't start the test server} );

    sleep 1; # background() may come back prematurely, so give it a second to fire up

    return $pid;
}

sub root {
    my $self = shift;
    my $port = $self->port;

    return "http://localhost:$port";
}

sub stop {
    if ( $pid ) {
        kill( 9, $pid ) unless $^S;
        undef $pid;
    }

    return;
}

1;


t/TestServer.pm
package TestServer;

use warnings;
use strict;

use Test::More;
use HTTP::Server::Simple::CGI;
use Cwd qw( realpath );
use parent qw( HTTP::Server::Simple::CGI );

my $dispatch_table = {};

=head1 OVERLOADED METHODS

=cut

our $pid;

sub new {
    die 'An instance of TestServer has already been started.' if $pid;

    my $class = shift;
    my $port  = shift;

    if ( !$port ) {
        $port = int(rand(20000)) + 20000;
    }
    my $self = $class->SUPER::new( $port );

    my $root = $self->root;

    return $self;
}

sub run {
    my $self = shift;

    $pid = $self->SUPER::run(@_);

    $SIG{__DIE__} = \&stop;

    return $pid;
}

sub handle_request {
    my $self = shift;
    my $cgi  = shift;

    my $path = $cgi->path_info();
    my $handler = $dispatch_table->{$path};

    if (ref($handler) eq 'CODE') {
        print "HTTP/1.0 200 OK\r\n";
        $handler->($cgi);
    }
    else {
        my $file = $path;
        if ( $file =~ m{/$} ) {
            $file .= 'index.html';
        }
        $file =~ s/\s+//g;
        my $filename = "t/html/$file";

        my ($code, $msg) = (200, 'OK');
        if ( ! -r $filename ) {
            ($code, $msg) = (404, 'Not Found');
        }
        if (index(realpath($filename), realpath("t/html")) != 0) {
            # don't expose a file outside server root
            ($code, $msg) = (403, 'Forbidden');
        }

        print "HTTP/1.0 $code $msg\r\n";

        if ($code == 200) {
            if (my $response=do { local (@ARGV, $/) = $filename; <> }) {
                print
                    "Content-Type: text/html\r\n",
                    "Content-Length: ", length($response), "\r\n\r\n",
                    $response;
            }
        }
        else {
            print
                $cgi->header,
                $cgi->start_html($msg),
                $cgi->h1($msg),
                $cgi->end_html;
        }
    }

    return;
}

=head1 METHODS UNIQUE TO TestServer

=cut

sub set_dispatch {
    my $self = shift;
    $dispatch_table = shift;

    return;
}

sub background {
    my $self = shift;

    $pid = $self->SUPER::background()
        or Carp::confess( q{Can't start the test server} );

    sleep 1; # background() may come back prematurely, so give it a second to fire up

    my $root = $self->root;

    diag( "Test server $root as PID $pid" );

    return $pid;
}


sub hostname {
    my $self = shift;

    return '127.0.0.1';
}

sub root {
    my $self = shift;
    my $port = $self->port;
    my $hostname = $self->hostname;

    return "http://$hostname:$port";
}

sub stop {
    if ( $pid ) {
        kill( 9, $pid ) unless $^S;
        undef $pid;
    }

    return;
}

1;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

my $parms = {
    NAME                => 'Test::WWW::Mechanize',
    AUTHOR              => 'Andy Lester <andy@petdance.com>',
    VERSION_FROM        => 'Mechanize.pm',
    ABSTRACT_FROM       => 'Mechanize.pm',
    PL_FILES            => {},
    PREREQ_PM           => {
        'Carp'                      => 0,
        'Carp::Assert::More'        => '1.16',
        'HTML::Form'                => 0,
        'HTML::TokeParser'          => 0,
        'HTTP::Message'             => '6.29',
        'LWP'                       => '6.02',
        'Test::Builder'             => 0,
        'Test::LongString'          => '0.15',
        'WWW::Mechanize'            => '1.68',
        'parent'                    => 0,
    },
    TEST_REQUIRES       => {
        'HTTP::Server::Simple'      => '0.42',
        'HTTP::Server::Simple::CGI' => 0,
        'Test::Builder::Tester'     => '1.09',
        'Test::More'                => '0.96', # subtest() and done_testing()
        'URI::file'                 => 0,
    },
    MIN_PERL_VERSION    => 5.010,
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Test-WWW-Mechanize-*' },
};

my $eumm_version = $ExtUtils::MakeMaker::VERSION;
$eumm_version =~ tr/_//d;

# TEST_REQUIRES is not supported in EUMM <= 6.63_03, so collapse the test
# prereqs into the build prereqs.
if ( $eumm_version <= 6.63_03 ) {
    $parms->{BUILD_REQUIRES} = {
        %{ $parms->{BUILD_REQUIRES} },
        %{ delete $parms->{TEST_REQUIRES} || {} },
    };
}
# BUILD_REQUIRES is not supported in EUMM <= 6.55_01, so collapse the build
# preereqs into the standard prereqs.
if ( $eumm_version <= 6.63_03 ) {
    $parms->{PREPREQ_PM} = {
        %{ $parms->{PREPREQ_PM} },
        %{ delete $parms->{BUILD_REQUIRES} || {} },
    };
}

if ( $eumm_version > 6.30 ) {
    $parms->{LICENSE} = 'artistic_2';
}

if ( $eumm_version >= 6.46 ) {
    $parms->{META_MERGE} = {
        resources => {
            license     => 'https://opensource.org/licenses/artistic-license-2.0',
            homepage    => 'https://github.com/petdance/test-www-mechanize',
            bugtracker  => 'https://github.com/petdance/test-www-mechanize/issues',
            repository  => 'https://github.com/petdance/test-www-mechanize',
        }
    };
}

my $module = 'HTML::Lint 2.20';
if ( not eval "use $module; 1;" ) {
    print "You don't have $module installed, so cannot use autolinting.\n";
}

WriteMakefile( %{$parms} );


sub MY::postamble {
    return <<'MAKE_FRAG';
.PHONY: critic tags

critic:
	perlcritic -1 -q -profile perlcriticrc Mechanize.pm t/

tags:
	ctags -f tags --recurse --totals \
		--exclude=blib \
		--exclude=.svn \
		--exclude='*~' \
		--languages=Perl --langmap=Perl:+.t \

MAKE_FRAG
}
